### PR TITLE
GH-35483: [CI][C++] Add header for snprintf for Windows

### LINKED ISSUE #35483: [CI][C++] conda-win-x64-cpu-py3 jobs failing due to snprintf definition
### Describe the bug, including details regarding any error messages, version, and platform.

https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=48306&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=41795ef0-6501-5db4-3ad4-33c0cf085626
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=48296&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=41795ef0-6501-5db4-3ad4-33c0cf085626

### Component(s)

C++, Continuous Integration

### PR DESCRIPTION
**Pull request by wjones127**
### Rationale for this change

Fixing failing nightlies.

### Are there any user-facing changes?

* Closes: #35483

### CODE REVIEW

**Comment by wjones127** [comment]
@github-actions crossbow submit conda-win-x64-cpu-py3

**Comment by wjones127** [comment]
@github-actions crossbow submit conda-win-x64-cpu-r41

File: python/pyarrow/src/arrow/python/platform.h [hunk 1]
```diff
@@ -32,5 +32,10 @@
 #if _MSC_VER >= 1900
 #undef timezone
 #endif
+
+// https://bugs.python.org/issue36020
+#if _MSC_VER >= 1500
+#undef snprintf
+#endif
```
Comment (kou) [inline python/pyarrow/src/arrow/python/platform.h:39]: It seems that the macro was removed by https://github.com/python/cpython/commit/e822e37946f27c09953bb5733acf3b07c2db690f .

How about the following?

```cpp
#ifdef snprintf
#undef snprintf
#endif
```

**Comment by kou** [comment]
@github-actions crossbow submit conda-win-*

**Review by kou — APPROVED** [review]
+1

The conda-win-x64-cpu-r41 failure is unrelated.

**Comment by kou** [comment]
Ah, I should have changed the PR title before I merge this...

**Pull request merged**